### PR TITLE
Get rid of LineTrackingStringBuffer class and instead use the line in…

### PR DESCRIPTION
**Pull request by ToddGrun**
…formation provided by RazorSourceDocument.

This additionally gets rid of an extra whole buffer allocation in the ParserContext. The most complex bit of the change is around avoiding TextLineCollection.GetLocation.

Overall, I'm seeing a pretty big win here, about 35% less time spent in RazorSyntaxTree.Parse for the typing scenario I was doing in a very large file.

### CODE REVIEW

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorSourceLineCollection.cs [hunk 1]
```diff
@@ -75,45 +75,31 @@ private int[] GetLineStarts()
...
+                        starts.Add(i + 1);
                         break;
 
                     case '\u0085':
                     case '\u2028':
                     case '\u2029':
-                        isLineBreak = true;
+                        starts.Add(i + 1);
                         break;
-
```
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorSourceLineCollection.cs:103]: 👏

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs [hunk 2]
```diff
@@ -3,32 +3,36 @@
...
+        public SeekableTextReader(RazorSourceDocument source)
         {
             if (source == null)
             {
                 throw new ArgumentNullException(nameof(source));
             }
 
-            _buffer = new LineTrackingStringBuffer(source, filePath);
+            _sourceDocument = source;
+            _cachedLineInfo = (new TextSpan(0, _sourceDocument.Lines.GetLineLength(0)), 0);
```
Comment (NTaylorMullen) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs:29]: This is confusing to me. Why cache the very first line and special case that?

Comment (ToddGrun) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs:29]: It simplified the code so it didnt need to check against uninitialized. Also, callers tend to start at the first line.

Comment (ajaybhargavb) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs:29]: This assumes that RazorSourceDocument will have atleast one line right? Is it safe to make that assumption?

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs [hunk 3]
```diff
@@ -55,22 +59,73 @@ public override int Read()
...
-            else if (_buffer.Length == 0)
+
+            if (_position < _sourceDocument.Length)
             {
-                _current = -1;
-                _location = SourceLocation.Zero;
+                if (_position >= _cachedLineInfo.Span.End)
+                {
+                    // Try to avoid the GetLocation call by checking if the next line contains the position
+                    int nextLineIndex = _cachedLineInfo.LineIndex + 1;
```
Comment (ajaybhargavb) [inline src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs:75]:
```suggestion
                    var nextLineIndex = _cachedLineInfo.LineIndex + 1;
```
Similarly in other places as well

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs (hunk 2 above)
Comment (ToddGrun) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs:29]: On empty documents, _sourceDocument.Lines.Count == 1. See [https://github.com/dotnet/aspnetcore/blob/66dd4919dafcb21c57a923a661cdbb0ff351713d/src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorSourceLineCollection.cs#L71](https://github.com/dotnet/aspnetcore/blob/66dd4919dafcb21c57a923a661cdbb0ff351713d/src/Razor/Microsoft.AspNetCore.Razor.Language/src/DefaultRazorSourceLineCollection.cs#L71)


---
In reply to: [440388134](https://github.com/dotnet/aspnetcore/pull/22903#discussion_r440388134) [](ancestors = 440388134)

File: src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs [hunk 4]
```diff
@@ -3,32 +3,36 @@
...
     {
-        private readonly LineTrackingStringBuffer _buffer;
+        private readonly RazorSourceDocument _sourceDocument;
         private int _position = 0;
         private int _current;
         private SourceLocation _location;
+        private (TextSpan Span, int LineIndex) _cachedLineInfo;
 
-        public SeekableTextReader(string source, string filePath) : this(source.ToCharArray(), filePath) { }
+        public SeekableTextReader(string source, string filePath) : this(new StringSourceDocument(source, Encoding.UTF8, new RazorSourceDocumentProperties(filePath, relativePath: null))) { }
```
[the remark this answers is not in the document]
Comment (ToddGrun) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs:19]: Two places from test code and once from LanguageCharacteristics.TokenizeString which is called from several places in product code. Perhaps UTF8 isn't appropriate here Encoding.Unicode?

---
In reply to: [440378902](https://github.com/dotnet/aspnetcore/pull/22903#discussion_r440378902) [](ancestors = 440378902)

[the remark this answers is not in the document]
Comment (ajaybhargavb) [reply src/Razor/Microsoft.AspNetCore.Razor.Language/src/Legacy/SeekableTextReader.cs:19]: Spoke offline. UTF8 is fine as we do the same thing in most other places as well.

**Review by ajaybhargavb — APPROVED** [review]

**Pull request merged**